format-parquet/src/main/java/io/cdap/plugin/format/parquet/input/ParquetInputFormatProvider.java

```diff
@@ -72,49 +75,53 @@ protected void addFormatProperties(Map<String, String> properties) {
 
   @Nullable
   @Override
-  public Schema getSchema(FormatContext context) {
-    if (conf.containsMacro(PathTrackingConfig.NAME_SCHEMA) || !Strings.isNullOrEmpty(conf.schema)) {
-      return super.getSchema(context);
+  public Schema detectSchema(FormatContext context, Supplier<InputStream> dataStream) throws IOException {
+    if (!(dataStream instanceof Function)) {
+      // should not happen with file source
+      return null;
     }
-    try {
-      return getDefaultSchema(context);
-    } catch (IOException e) {
-      throw new IllegalArgumentException("Invalid schema: " + e.getMessage(), e);
+
+    //noinspection unchecked
+    long len = ((Function<Void, Long>) dataStream).apply(null);
+    InputFile hadoopInputFile = new HadoopInputFile(len, dataStream);
+    ParquetReadOptions readOptions = ParquetReadOptions.builder().build();
+    try (ParquetFileReader parquetFileReader = ParquetFileReader.open(hadoopInputFile, readOptions)) {
+      MessageType parquetSchema = parquetFileReader.getFooter().getFileMetaData().getSchema();
+      return new AvroToStructuredTransformer().convertSchema(new AvroSchemaConverter().convert(parquetSchema));
     }
   }
 
-  /**
-   * Extract schema from file
-   *
-   * @param context {@link FormatContext}
-   * @return {@link Schema}
-   * @throws IOException raised when error occurs during schema extraction
-   */
-  public Schema getDefaultSchema(FormatContext context) throws IOException {
-    String filePath = conf.getProperties().getProperties().getOrDefault("path", null);
-    Job job = JobUtils.createInstance();
-    Configuration hconf = job.getConfiguration();
+  private static class HadoopInputFile implements InputFile {
+    private final long length;
+    private final Supplier<InputStream> streamSupplier;
+
+    HadoopInputFile(long length, Supplier<InputStream> streamSupplier) {
+      this.length = length;
+      this.streamSupplier = streamSupplier;
+    }
 
-    for (Map.Entry<String, String> entry : conf.getFileSystemProperties().entrySet()) {
-      hconf.set(entry.getKey(), entry.getValue());
+    @Override
+    public long getLength() {
+      return length;
     }
 
-    List<Path> paths = conf.getFilePathsForSchemaGeneration(filePath, ".+\\.parquet$", hconf, job);
-    for (Path file : paths) {
-      try (ParquetReader reader = AvroParquetReader.builder(file).withConf(hconf).build()) {
-        GenericData.Record record = (GenericData.Record) reader.read();
-        if (record == null) {
-          continue;
+    @Override
+    public SeekableInputStream newStream() {
+      // hack, ValidatingInputFormat.detectSchema() should be updated to provide a seekable stream
```

Comment:
Any reason we can't do that?

Reply:
I need to cherry pick this onto the release branch, so can't update the API there. Can make the change in CDAP afterwards.

Reply:
Though I'm thinking it may actually be better to just add in in 6.7.1 since nothing was using it in 6.7.0. Would avoid a bunch of complications around compatibility